### PR TITLE
Refactor the KeyLIght properties in their own group

### PR DESCRIPTION
**Pull request by samcake**
And  add an AmbientURL property to later add a way to specify the ambient component per zone from an url: a cubemap

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: bc7ad9d3f78570b02dff9e9718897181b5f8b2f2 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

File: libraries/entities/src/KeyLightPropertyGroup.h [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+class EntityItemProperties;
+class EncodeBitstreamParams;
+class OctreePacketData;
+class EntityTreeElementExtraEncodeData;
+class ReadBitstreamToTreeParams;
+
+class KeyLightPropertyGroup : public PropertyGroup {
+public:
+    //void associateWithAnimationLoop(AnimationLoop* animationLoop) { _animationLoop = animationLoop; }
```
Comment (ZappoMan) [inline libraries/entities/src/KeyLightPropertyGroup.h]: dead code

File: libraries/entities/src/KeyLightPropertyGroup.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    static const glm::vec3 DEFAULT_KEYLIGHT_DIRECTION;
+    
+    DEFINE_PROPERTY_REF(PROP_KEYLIGHT_COLOR, Color, color, xColor, DEFAULT_KEYLIGHT_COLOR);
+    DEFINE_PROPERTY(PROP_KEYLIGHT_INTENSITY, Intensity, intensity, float, DEFAULT_KEYLIGHT_INTENSITY);
+    DEFINE_PROPERTY(PROP_KEYLIGHT_AMBIENT_INTENSITY, AmbientIntensity, ambientIntensity, float, DEFAULT_KEYLIGHT_AMBIENT_INTENSITY);
+    DEFINE_PROPERTY_REF(PROP_KEYLIGHT_DIRECTION, Direction, direction, glm::vec3, DEFAULT_KEYLIGHT_DIRECTION);
+    DEFINE_PROPERTY_REF(PROP_KEYLIGHT_AMBIENT_URL, AmbientURL, ambientURL, QString, "");
+    
+    
+    /*DEFINE_PROPERTY_REF(PROP_ANIMATION_URL, URL, url, QString, "");
```
Comment (ZappoMan) [inline libraries/entities/src/KeyLightPropertyGroup.h]: dead code

File: libraries/entities/src/KeyLightPropertyGroup.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+     
+     float getKeyLightAmbientIntensity() const { return _keyLightAmbientIntensity; }
+     void setKeyLightAmbientIntensity(float value) { _keyLightAmbientIntensity = value; }
+     
+     const glm::vec3& getKeyLightDirection() const { return _keyLightDirection; }
+     void setKeyLightDirection(const glm::vec3& value) { _keyLightDirection = value; }
+     */
+    
+protected:
+   // void setFromOldAnimationSettings(const QString& value);
```
Comment (ZappoMan) [inline libraries/entities/src/KeyLightPropertyGroup.h]: dead code

File: libraries/entities/src/KeyLightPropertyGroup.h [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+     _keyLightColor[BLUE_INDEX] = value.blue;
+     }
+     
+     void setKeyLightColor(const rgbColor& value) {
+     _keyLightColor[RED_INDEX] = value[RED_INDEX];
+     _keyLightColor[GREEN_INDEX] = value[GREEN_INDEX];
+     _keyLightColor[BLUE_INDEX] = value[BLUE_INDEX];
+     }
+     
+     glm::vec3 getKeyLightColorVec3() const {
```
Comment (ZappoMan) [inline libraries/entities/src/KeyLightPropertyGroup.h]: this one might be useful to keep

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 5]
```diff
@@ -270,10 +270,11 @@ void EntityTreeRenderer::applyZonePropertiesToScene(std::shared_ptr<ZoneEntityIt
             _previousStageDay = scene->getStageYearTime();
             _hasPreviousZone = true;
         }
-        scene->setKeyLightColor(zone->getKeyLightColorVec3());
-        scene->setKeyLightIntensity(zone->getKeyLightIntensity());
-        scene->setKeyLightAmbientIntensity(zone->getKeyLightAmbientIntensity());
-        scene->setKeyLightDirection(zone->getKeyLightDirection());
+        auto xcolor = zone->getKeyLightProperties().getColor();
+        scene->setKeyLightColor(glm::vec3(xcolor.red / 256.0f, xcolor.green / 256.0f, xcolor.blue / 256.0f));
```
Comment (ZappoMan) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: 1) you might want to add back... getKeyLightColorVec3() so you can do this in the property group.
2) magic numbers

**Comment by hifi-gustavo** [comment]
Build console output for commit: ca30ce1ff0fc62e8ebc50df5585d92285a929a88 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ff57b73bd6293697c77198c8cc4531a0a4d5e1ef available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

File: libraries/shared/src/ColorUtils.h [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "DependencyManager.h"
+
+class ColorUtils {
+public:
+    inline static glm::vec3 toVec3(const xColor& color);
+};
+
+inline glm::vec3 ColorUtils::toVec3(const xColor& color) {
+    const float ONE_OVER_256 = 1.0f / 256.0f;
```
Comment (ZappoMan) [inline libraries/shared/src/ColorUtils.h]: shouldn't this be 255? isn't the max 8bit color value actually 255... and so with this code you can never actually get to 1.0f?

Comment (samcake) [inline libraries/shared/src/ColorUtils.h]: there  is good
and there is brad's good

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8e2c269cc0c66a57f2a1f86acc5c82c3abab2442 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by ZappoMan** [comment]
I've reviewed this and it's ready to merge once we're done with our demos.

**Comment by samcake** [comment]
I also need to fix the edit.js to reflect that change properly n the ui too.

**Comment by ZappoMan** [comment]
good point @samcake - we should do that before we merge

**Comment by hifi-gustavo** [comment]
Build console output for commit: e4aab4bd97b0c080af9a97b3925450e7794eabc9 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a5b95d73073ec1ebae2780eb2bf15d071d2e3f69 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 630cdb8a9dd38e4d2ca9cacbf727101ecd5d3625 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fd7582ad31e3b17c30628455abd7f6ba62b63df8 available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by ZappoMan** [comment]
In testing I found one bug - It seems like the keyLightDirection is not being "preserved" when loading old model.json.gz files. I had a zone with a keyLightDirection set to (45, 45, 0). But when I opened up in the new server, the new direction was something like (0.5, -0.7, 0.5) -- it looks like this is a radians to degrees conversion issue.

**Comment by ZappoMan** [comment]
I think the problem is in this code in edit.js....

line 1554... you didn't change that to look at the new `keylight.direction` property
```
if (data.properties.keyLightDirection !== undefined) {
    data.properties.keyLightDirection = Vec3.fromPolar(
                        data.properties.keyLightDirection.x * DEGREES_TO_RADIANS, data.properties.keyLightDirection.y * DEGREES_TO_RADIANS);
}

```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 025422792d134fefd4fe65a4815f6c5197849f4e available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 517387621a5c0b5389d240bb8846dbef7bd0421c available at: https://gist.github.com/b483f923f8eb7a5e21f0<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6178.dmg

**Pull request merged**